README
NAME
    ZeroMQ - A ZeroMQ2 wrapper for Perl

SYNOPSIS ( HIGH-LEVEL API )
        # echo server
        use ZeroMQ qw/:all/;

        my $cxt = ZeroMQ::Context->new;
        my $sock = $cxt->socket(ZMQ_REP);
        $sock->bind($addr);
  
        my $msg;
        foreach (1..$roundtrip_count) {
            $msg = $sock->recv();
            $sock->send($msg);
        }

        # json (if JSON.pm is available)
        $sock->send_as( json => { foo => "bar" } );
        my $thing = $sock->recv_as( "json" );

        # custom serialization
        ZeroMQ::register_read_type(myformat => sub { ... });
        ZeroMQ::register_write_type(myformat => sub { .. });

        $sock->send_as( myformat => $data ); # serialize using above callback
        my $thing = $sock->recv_as( "myformat" );

SYNOPSIS ( LOW-LEVEL API )
        use ZeroMQ::Raw;

        my $ctxt = zmq_init($threads);
        my $rv   = zmq_term($ctxt);

        my $msg  = zmq_msg_init();
        my $msg  = zmq_msg_init_size( $size );
        my $msg  = zmq_msg_init_data( $data );
        my $rv   = zmq_msg_close( $msg );
        my $rv   = zmq_msg_move( $dest, $src );
        my $rv   = zmq_msg_copy( $dest, $src );
        my $data = zmq_msg_data( $msg );
        my $size = zmq_msg_size( $msg);

        my $sock = zmq_socket( $ctxt, $type );
        my $rv   = zmq_close( $sock );
        my $rv   = zmq_setsockopt( $socket, $option, $value );
        my $val  = zmq_getsockopt( $socket, $option );
        my $rv   = zmq_bind( $sock, $addr );
        my $rv   = zmq_send( $sock, $msg, $flags );
        my $msg  = zmq_recv( $sock, $flags );

INSTALLATION
    If you have libzmq registered with pkg-config:

        perl Makefile.PL
        make 
        make test
        make install

    If you don't have pkg-config, and libzmq is installed under
    /usr/local/libzmq:

        ZMQ_HOME=/usr/local/libzmq \
            perl Makefile.PL
        make
        make test
        make install

    If you want to customize include directories and such:

        ZMQ_INCLUDES=/path/to/libzmq/include \
        ZMQ_LIBS=/path/to/libzmq/lib \
        ZMQ_H=/path/to/libzmq/include/zmq.h \
            perl Makefile.PL
        make
        make test
        make install

    If you want to compile with debugging on:

        perl Makefile.PL -g

DESCRIPTION
    The "ZeroMQ" module is a wrapper of the 0MQ message passing library for
    Perl. It's a thin wrapper around the C API. Please read
    <http://zeromq.org> for more details on ZeroMQ.

CLASS WALKTHROUGH
    ZeroMQ::Raw
        Use ZeroMQ::Raw to get access to the C API such as "zmq_init",
        "zmq_socket", et al. Functions provided in this low level API should
        follow the C API exactly.

    ZeroMQ::Constants
        ZeroMQ::Constants contains all of the constants that are known to be
        extractable from zmq.h. Do note that sometimes the list changes due
        to additions/deprecations in the underlying zeromq2 library. We try
        to do our best to make things available (at least to warn you that
        some symbols are deprecated), but it may not always be possible.

    ZeroMQ::Context
    ZeroMQ::Socket
    ZeroMQ::Message
        ZeroMQ::Context, ZeroMQ::Socket, ZeroMQ::Message contain the
        high-level, more perl-ish interface to the zeromq functionalities.

    ZeroMQ
        Loading "ZeroMQ" will make the ZeroMQ::Context, ZeroMQ::Socket, and
        ZeroMQ::Message classes available as well.

BASIC USAGE
    To start using ZeroMQ, you need to create a context object, then as many
    ZeroMQ::Socket as you need:

        my $ctxt = ZeroMQ::Context->new;
        my $socket = $ctxt->socket( ... options );

    You need to call "bind()" or "connect()" on the socket, depending on
    your usage. For example on a typical server-client model you would write
    on the server side:

        $socket->bind( "tcp://127.0.0.1:9999" );

    and on the client side:

        $socket->connect( "tcp://127.0.0.1:9999" );

    The underlying zeromq library offers TCP, multicast, in-process, and ipc
    connection patterns. Read the zeromq manual for more details on other
    ways to setup the socket.

    When sending data, you can either pass a ZeroMQ::Message object or a
    Perl string.

        # the following two send() calls are equivalent
        my $msg = ZeroMQ::Message->new( "a simple message" );
        $socket->send( $msg );
        $socket->send( "a simple message" );

    In most cases using ZeroMQ::Message is redundunt, so you will most
    likely use the string version.

    To receive, simply call "recv()" on the socket

        my $msg = $socket->recv;

    The received message is an instance of ZeroMQ::Message object, and you
    can access the content held in the message via the "data()" method:

        my $data = $msg->data;

SERIALIZATION
    ZeroMQ.pm comes with a simple serialization/deserialization mechanism.

    To serialize, use "register_write_type()" to register a name and an
    associated callback to serialize the data. For example, for JSON we do
    the following (this is already done for you in ZeroMQ.pm if you have
    JSON.pm installed):

        use JSON ();
        ZeroMQ::register_write_type('json' => \&JSON::encode_json);
        ZeroMQ::register_read_type('json' => \&JSON::decode_json);

    Then you can use "send_as()" and "recv_as()" to specify the
    serialization type as the first argument:

        my $ctxt = ZeroMQ::Context->new();
        my $sock = $ctxt->socket( ZMQ_REQ );

        $sock->send_as( json => $complex_perl_data_structure );

    The otherside will receive a JSON encoded data. The receivind side can
    be written as:

        my $ctxt = ZeroMQ::Context->new();
        my $sock = $ctxt->socket( ZMQ_REP );

        my $complex_perl_data_structure = $sock->recv_as( 'json' );

    If you have JSON.pm (must be 2.00 or above), then the JSON serializer /
    deserializer is automatically enabled. If you want to tweak the
    serializer option, do something like this:

        my $coder = JSON->new->utf8->pretty; # pretty print
        ZeroMQ::register_write_type( json => sub { $coder->encode($_[0]) } );
        ZeroMQ::register_read_type( json => sub { $coder->decode($_[0]) } );

    Note that this will have a GLOBAL effect. If you want to change only
    your application, use a name that's different from 'json'.

ASYNCHRONOUS I/O WITH ZEROMQ
    By default ZeroMQ comes with its own zmq_poll() mechanism that can
    handle non-blocking sockets. You can use this by calling zmq_poll with a
    list of hashrefs:

        zmq_poll([
            {
                fd => fileno(STDOUT),
                events => ZMQ_POLLOUT,
                callback => \&callback,
            },
            {
                socket => $zmq_socket,
                events => ZMQ_POLLIN,
                callback => \&callback
            },
        ], $timeout );

    Unfortunately this custom polling scheme doesn't play too well with
    AnyEvent.

    As of zeromq2-2.1.0, you can use getsockopt to retrieve the underlying
    file descriptor, so use that to integrate ZeroMQ and AnyEvent:

        my $socket = zmq_socket( $ctxt, ZMQ_REP );
        my $fh = zmq_getsockopt( $socket, ZMQ_FD );
        my $w; $w = AE::io $fh, 0, sub {
            while ( my $msg = zmq_recv( $socket, ZMQ_RCVMORE ) ) {
                # do something with $msg;
            }
            undef $w;
        };

NOTES ON MULTI-PROCESS and MULTI-THREADED USAGE
    ZeroMQ works on both multi-process and multi-threaded use cases, but you
    need to be careful bout sharing ZeroMQ objects.

    For multi-process environments, you should not be sharing the context
    object. Create separate contexts for each process, and therefore you
    shouldn't be sharing the socket objects either.

    For multi-thread environemnts, you can share the same context object.
    However you cannot share sockets.

FUNCTIONS
  version()
    Returns the version of the underlying zeromq library that is being
    linked. In scalar context, returns a dotted version string. In list
    context, returns a 3-element list of the version numbers:

        my $version_string = ZeroMQ::version();
        my ($major, $minor, $patch) = ZeroMQ::version();

  device($type, $sock1, $sock2)
  register_read_type($name, \&callback)
    Register a read callback for a given $name. This is used in "recv_as()".
    The callback receives the data received from the socket.

  register_write_type($name, \&callback)
    Register a write callback for a given $name. This is used in "send_as()"
    The callback receives the Perl structure given to "send_as()"

DEBUGGING XS
    If you see segmentation faults, and such, you need to figure out where
    the error is occuring in order for the maintainers to figure out what
    happened. Here's a very very brief explanation of steps involved.

    First, make sure to compile ZeroMQ.pm with debugging on by specifying
    -g:

        perl Makefile.PL -g
        make

    Then fire gdb:

        gdb perl
        (gdb) R -Mblib /path/to/your/script.pl

    When you see the crash, get a backtrace:

        (gdb) bt

CAVEATS
    This is an early release. Proceed with caution, please report (or better
    yet: fix) bugs you encounter.

    This module has been tested againt zeromq 2.1.4. Semantics of this
    module rely heavily on the underlying zeromq version. Make sure you know
    which version of zeromq you're working with.

SEE ALSO
    ZeroMQ::Raw, ZeroMQ::Context, ZeroMQ::Socket, ZeroMQ::Message

    <http://zeromq.org>

    <http://github.com/lestrrat/ZeroMQ-Perl>

AUTHOR
    Daisuke Maki "<daisuke@endeworks.jp>"

    Steffen Mueller, "<smueller@cpan.org>"

COPYRIGHT AND LICENSE
    The ZeroMQ module is

    Copyright (C) 2010 by Daisuke Maki

    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself, either Perl version 5.8.0 or, at
    your option, any later version of Perl 5 you may have available.



Changes
Changelog for Perl module ZeroMQ.

0.21 Mar 22 2012
  - Fix threading issues causing a segfault.
    martyn-gillmore/aero did the diagnosis on Win32, but that fix still
    caused problems in *nix, which showed a deeper problem. rafl provided
    a new patch, and hopefully this will fix the problem. 
        * Initial fix by martyn-gillmore/aero:
            https://github.com/lestrrat/ZeroMQ-Perl/commit/baeb5233749233ced890141f5ab961fea1b1055d
        * Additional fix by rafl:
            https://github.com/lestrrat/ZeroMQ-Perl/commit/c9c7e0f1b98d7fc8336a612aa0b15f87964b3ef0
  - Doc tweaks (smueller)
  - Add ROUTER/DEALER
  - HEADS UP: There are plans to ditch ZeroMQ.pm in favor of multiple
    implementations, to provide more stability:
        * https://github.com/lestrrat/p5-ZMQ/
        * https://github.com/lestrrat/ZeroMQ-Perl/issues/36
        * https://github.com/lestrrat/ZeroMQ-Perl/issues/35

0.20 Jan 12 2012
  - Fix recv_as to return undef when ZMQ_NOBLOCK is specified
    (https://github.com/lestrrat/ZeroMQ-Perl/pull/31 by sebnow)
  - Make sure to not double-free C structures when explicit closing
    has been performed
  - Add ZMQ_RECONNECT_IVL_MAX in get/setsopckopt
  - Tweak automated tests

0.19 Dec 08 2011
  - Fix socket/context destruction order (github #20).
  - Apply doc patches.
  - Make ZMQ_NOBLOCK to ZMQ_DONTWAIT when libzmq >= 3
  - Change tests to using Test::Fatal instead of Test::Exception
  - Change tests to using Test::TCP object interface

0.18 Nov 06 2011
  - Pass $flags in ZeroMQ::Socket->recv_as() as is documented

0.17 Fri Sep 09 2011 12:40 JST
  - Add support for ZMQ_LINGER in setsockopt (sebnow)
    https://github.com/lestrrat/ZeroMQ-Perl/pull/19

0.16 Sat Jul 23 2011 08:00 JST
  - Fix memory leaks
    * rt #69572
    * https://github.com/lestrrat/ZeroMQ-Perl/pull/17 (linuxfood)
  - Make sure to check for libzmq 2.1.x, as we don't support anything else
    on this release

0.15 Mon May 23 2011 17:30 JST
  - Do not use PKG_CONFIG_PATH as the path to pkg-config binary.
    If for some reason you want to use a different pkg-config than
    in your PATH, use PKGCONFIG_CMD environmental variable.
    Reported by Johan Ström (rt #68108)
  - Doc tweaks
  - Fix/update autobuild tools used in Jenkins tests.

0.14 Thu May 12 2011 17:05 JST
  - Argh, forgot to change all the occurances of pkg-config to $pkg_config

0.13 Thu May 12 2011 17:00 JST
  - Fix problems introduced by using Devel::CheckLib + ZMQ_HOME et al.
    Problem reported by Johan Ström (rt #68108)
  - Respect PKG_CONFIG_PATH when probing for zeromq

0.12 Fri Apr 15 2011 15:35 JST
  - Add missing tools/detect_zmq.pl which implemented the detection
    scheme introduced in 0.11

0.11 Fri Apr 15 2011 15:30 JST
  - Make header, incpath, lib detection smarter. 
    * Now supports pkg-config!
    * Environment variables ZMQ_HOME, ZMQ_INCLUDES, ZMQ_LIBS are
      honored. Please see POD for how to configure ZeroMQ.pm
  - Include a section on how to install in the POD.
  - Remove usage of DOWNSTREAM/UPSTREAM, and connect inproc
    sockets using PAIR

0.10 Wed Apr 13 2011 10:30 JST
  - Use Module::Instsall::CheckLib
  - Silence test failures [https://github.com/lestrrat/ZeroMQ-Perl/issues/11]
    (reported by oyvindsk)

0.09_01 Fri Mar 11 2011 08:00 JST
  - Add ZeroMQ::Poller (https://github.com/lestrrat/ZeroMQ-Perl/pull/8)

0.09 Wed Feb 16 2011 22:30 JST
  - Only run silly Module::Install hoopla if you're an author
  - Tweak Makefile.PL
  - Smarter zmq.h detection
  - Include Devel::CheckLib in requires

0.08 Sun Jan 23 2011 14:45 JST
  - Fixed many memory leaks (rt #64944) reported by Jason Ball
    * Plugged ZeroMQ::Raw::Message related leaks 
    * Plugged leaks when setting $!
    * Wrote tests using zmq_poll() and AnyEvent to do non-blocking
      recv() with zeromq
  - Backwards incompatible change! zmq_send() now returns exactly what
    the underlying zeromq library returns -- previously it used to
    return a "true" value for success and false otherwise.

0.07 Tue Jan 18 2011 07:50 JST
  - Somehow zmq_device() was missing through 0.03 to 0.06. Re-added
    as ZeroMQ::Raw::zmq_device()

0.06 Tue Jan 18 2011 07:20 JST
  - Fix critical message corruption bug introduced by wrong usage of
    zmq_msg_init_data() (rt #64836). Reported, diagnosed by Jason Ball

0.05 Thu Jan 06 2011 19:20 JST
  - Various issues (rt #64520) reported by Jason Ball
    * Make sure to generate const-xs.inc and the like when installing
    * Tweak to work with perls older than 5.10
    * Only generate README for authors

0.04 Wed Jan 05 2011 18:20 JST
  - Grr, remove autobuild stuff from MANIFEST, repackage.
  - Code is identical to 0.03

0.03 Wed Jan 05 2011 15:00 JST
  - Ad docs, and upload 0.03

0.02_05 Tue Jan 03 2011 00:30 JST
  - Add zmq_poll, with the ability to poll random file descriptors along with
    zmq sockets (rt #62851)

0.02_04 Tue Dec 29 2010 16:20 JST
  - Add zmq_msg_copy and zmq_msg_move
  - Fix minor compile warning
  - Make it so that ZeroMQ::Raw doesn't depend on ZeroMQ -- and other
    modules that it in turn depends on. Now ZeroMQ::Raw can be used standalone

0.02_03 Tue Dec 28 2010 21:40 JST
  - Rework the internals to allow high level API (perl-ish, OO) and
    low level API (straight portof zeromq C binding), thanks to jrockway
  - Remove PollItem in favor of getsockopt + ZMQ_FD. Holler if you need it back

0.02_02 Sat Oct 30 2010 21:20 JST
  - Update to work with zeromq2 master (as of 10/28)
  - Added new constants
  - ZMQ_POLLIN callbacks used to pass the ZeroMQ::Message argument
    to the callback function, but this is error prone, so now the callback
    does not receive anything

0.02_01 Mon Sep 06 2010 19:20 JST
  - Releasing a dev release to get some test coverage
  - Add zmq_poll support via ZeroMQ::PollItem
  - Lots of doc fixes
  - Lots of test fixes
  - For people cloning from github, tell them what Module::Install
    modules you need to install

lib/ZeroMQ.pm
package ZeroMQ;
use strict;
BEGIN {
    our $VERSION = '0.21';
    our @ISA = qw(Exporter);
}
use ZeroMQ::Raw ();
use ZeroMQ::Context;
use ZeroMQ::Socket;
use ZeroMQ::Message;
use ZeroMQ::Poller;
use ZeroMQ::Constants;
use 5.008;
use Carp ();
use IO::Handle;

our %SERIALIZERS;
our %DESERIALIZERS;
sub register_read_type { $DESERIALIZERS{$_[0]} = $_[1] }
sub register_write_type { $SERIALIZERS{$_[0]} = $_[1] }

sub import {
    my $class = shift;
    if (@_) {
        ZeroMQ::Constants->export_to_level( 1, $class, @_ );
    }
}

sub _get_serializer { $SERIALIZERS{$_[1]} }
sub _get_deserializer { $DESERIALIZERS{$_[1]} }

eval {
    require JSON;
    JSON->import(2.00);
    register_read_type(json => \&JSON::decode_json);
    register_write_type(json => \&JSON::encode_json);
};

1;
__END__

=head1 NAME

ZeroMQ - A ZeroMQ2 wrapper for Perl

=head1 SYNOPSIS ( HIGH-LEVEL API )

    # echo server
    use ZeroMQ qw/:all/;

    my $cxt = ZeroMQ::Context->new;
    my $sock = $cxt->socket(ZMQ_REP);
    $sock->bind($addr);
  
    my $msg;
    foreach (1..$roundtrip_count) {
        $msg = $sock->recv();
        $sock->send($msg);
    }

    # json (if JSON.pm is available)
    $sock->send_as( json => { foo => "bar" } );
    my $thing = $sock->recv_as( "json" );

    # custom serialization
    ZeroMQ::register_read_type(myformat => sub { ... });
    ZeroMQ::register_write_type(myformat => sub { .. });

    $sock->send_as( myformat => $data ); # serialize using above callback
    my $thing = $sock->recv_as( "myformat" );

=head1 SYNOPSIS ( LOW-LEVEL API )

    use ZeroMQ::Raw;

    my $ctxt = zmq_init($threads);
    my $rv   = zmq_term($ctxt);

    my $msg  = zmq_msg_init();
    my $msg  = zmq_msg_init_size( $size );
    my $msg  = zmq_msg_init_data( $data );
    my $rv   = zmq_msg_close( $msg );
    my $rv   = zmq_msg_move( $dest, $src );
    my $rv   = zmq_msg_copy( $dest, $src );
    my $data = zmq_msg_data( $msg );
    my $size = zmq_msg_size( $msg);

    my $sock = zmq_socket( $ctxt, $type );
    my $rv   = zmq_close( $sock );
    my $rv   = zmq_setsockopt( $socket, $option, $value );
    my $val  = zmq_getsockopt( $socket, $option );
    my $rv   = zmq_bind( $sock, $addr );
    my $rv   = zmq_send( $sock, $msg, $flags );
    my $msg  = zmq_recv( $sock, $flags );

=head1 INSTALLATION

If you have libzmq registered with pkg-config:

    perl Makefile.PL
    make 
    make test
    make install

If you don't have pkg-config, and libzmq is installed under /usr/local/libzmq:

    ZMQ_HOME=/usr/local/libzmq \
        perl Makefile.PL
    make
    make test
    make install

If you want to customize include directories and such:

    ZMQ_INCLUDES=/path/to/libzmq/include \
    ZMQ_LIBS=/path/to/libzmq/lib \
    ZMQ_H=/path/to/libzmq/include/zmq.h \
        perl Makefile.PL
    make
    make test
    make install

If you want to compile with debugging on:

    perl Makefile.PL -g

=head1 DESCRIPTION

The C<ZeroMQ> module is a wrapper of the 0MQ message passing library for Perl. 
It's a thin wrapper around the C API. Please read L<http://zeromq.org> for
more details on ZeroMQ.

=head1 CLASS WALKTHROUGH

=over 4

=item ZeroMQ::Raw

Use L<ZeroMQ::Raw> to get access to the C API such as C<zmq_init>, C<zmq_socket>, et al. Functions provided in this low level API should follow the C API exactly.

=item ZeroMQ::Constants

L<ZeroMQ::Constants> contains all of the constants that are known to be extractable from zmq.h. Do note that sometimes the list changes due to additions/deprecations in the underlying zeromq2 library. We try to do our best to make things available (at least to warn you that some symbols are deprecated), but it may not always be possible.

=item ZeroMQ::Context

=item ZeroMQ::Socket

=item ZeroMQ::Message

L<ZeroMQ::Context>, L<ZeroMQ::Socket>, L<ZeroMQ::Message> contain the high-level, more perl-ish interface to the zeromq functionalities.

=item ZeroMQ

Loading C<ZeroMQ> will make the L<ZeroMQ::Context>, L<ZeroMQ::Socket>, and 
L<ZeroMQ::Message> classes available as well.

=back

=head1 BASIC USAGE

To start using ZeroMQ, you need to create a context object, then as many ZeroMQ::Socket as you need:

    my $ctxt = ZeroMQ::Context->new;
    my $socket = $ctxt->socket( ... options );

You need to call C<bind()> or C<connect()> on the socket, depending on your usage. For example on a typical server-client model you would write on the server side:

    $socket->bind( "tcp://127.0.0.1:9999" );

and on the client side:

    $socket->connect( "tcp://127.0.0.1:9999" );

The underlying zeromq library offers TCP, multicast, in-process, and ipc connection patterns. Read the zeromq manual for more details on other ways to setup the socket.

When sending data, you can either pass a ZeroMQ::Message object or a Perl string. 

    # the following two send() calls are equivalent
    my $msg = ZeroMQ::Message->new( "a simple message" );
    $socket->send( $msg );
    $socket->send( "a simple message" ); 

In most cases using ZeroMQ::Message is redundunt, so you will most likely use the string version.

To receive, simply call C<recv()> on the socket

    my $msg = $socket->recv;

The received message is an instance of ZeroMQ::Message object, and you can access the content held in the message via the C<data()> method:

    my $data = $msg->data;

=head1 SERIALIZATION

ZeroMQ.pm comes with a simple serialization/deserialization mechanism.

To serialize, use C<register_write_type()> to register a name and an
associated callback to serialize the data. For example, for JSON we do
the following (this is already done for you in ZeroMQ.pm if you have
JSON.pm installed):

    use JSON ();
    ZeroMQ::register_write_type('json' => \&JSON::encode_json);
    ZeroMQ::register_read_type('json' => \&JSON::decode_json);

Then you can use C<send_as()> and C<recv_as()> to specify the serialization 
type as the first argument:

    my $ctxt = ZeroMQ::Context->new();
    my $sock = $ctxt->socket( ZMQ_REQ );

    $sock->send_as( json => $complex_perl_data_structure );

The otherside will receive a JSON encoded data. The receivind side
can be written as:

    my $ctxt = ZeroMQ::Context->new();
    my $sock = $ctxt->socket( ZMQ_REP );

    my $complex_perl_data_structure = $sock->recv_as( 'json' );

If you have JSON.pm (must be 2.00 or above), then the JSON serializer / 
deserializer is automatically enabled. If you want to tweak the serializer
option, do something like this:

    my $coder = JSON->new->utf8->pretty; # pretty print
    ZeroMQ::register_write_type( json => sub { $coder->encode($_[0]) } );
    ZeroMQ::register_read_type( json => sub { $coder->decode($_[0]) } );

Note that this will have a GLOBAL effect. If you want to change only
your application, use a name that's different from 'json'.

=head1 ASYNCHRONOUS I/O WITH ZEROMQ

By default ZeroMQ comes with its own zmq_poll() mechanism that can handle
non-blocking sockets. You can use this by calling zmq_poll with a list of
hashrefs:

    zmq_poll([
        {
            fd => fileno(STDOUT),
            events => ZMQ_POLLOUT,
            callback => \&callback,
        },
        {
            socket => $zmq_socket,
            events => ZMQ_POLLIN,
            callback => \&callback
        },
    ], $timeout );

Unfortunately this custom polling scheme doesn't play too well with AnyEvent.

As of zeromq2-2.1.0, you can use getsockopt to retrieve the underlying file
descriptor, so use that to integrate ZeroMQ and AnyEvent:

    my $socket = zmq_socket( $ctxt, ZMQ_REP );
    my $fh = zmq_getsockopt( $socket, ZMQ_FD );
    my $w; $w = AE::io $fh, 0, sub {
        while ( my $msg = zmq_recv( $socket, ZMQ_RCVMORE ) ) {
            # do something with $msg;
        }
        undef $w;
    };

=head1 NOTES ON MULTI-PROCESS and MULTI-THREADED USAGE

ZeroMQ works on both multi-process and multi-threaded use cases, but you need
to be careful bout sharing ZeroMQ objects.

For multi-process environments, you should not be sharing the context object.
Create separate contexts for each process, and therefore you shouldn't
be sharing the socket objects either.

For multi-thread environemnts, you can share the same context object. However
you cannot share sockets.

=head1 FUNCTIONS

=head2 version()

Returns the version of the underlying zeromq library that is being linked.
In scalar context, returns a dotted version string. In list context,
returns a 3-element list of the version numbers:

    my $version_string = ZeroMQ::version();
    my ($major, $minor, $patch) = ZeroMQ::version();

=head2 device($type, $sock1, $sock2)

=head2 register_read_type($name, \&callback)

Register a read callback for a given C<$name>. This is used in C<recv_as()>.
The callback receives the data received from the socket.

=head2 register_write_type($name, \&callback)

Register a write callback for a given C<$name>. This is used in C<send_as()>
The callback receives the Perl structure given to C<send_as()>

=head1 DEBUGGING XS

If you see segmentation faults, and such, you need to figure out where the error is occuring in order for the maintainers to figure out what happened. Here's a very very brief explanation of steps involved.

First, make sure to compile ZeroMQ.pm with debugging on by specifying -g:

    perl Makefile.PL -g
    make

Then fire gdb:

    gdb perl
    (gdb) R -Mblib /path/to/your/script.pl

When you see the crash, get a backtrace:

    (gdb) bt

=head1 CAVEATS

This is an early release. Proceed with caution, please report
(or better yet: fix) bugs you encounter.

This module has been tested againt B<zeromq 2.1.4>. Semantics of this
module rely heavily on the underlying zeromq version. Make sure
you know which version of zeromq you're working with.

=head1 SEE ALSO

L<ZeroMQ::Raw>, L<ZeroMQ::Context>, L<ZeroMQ::Socket>, L<ZeroMQ::Message>

L<http://zeromq.org>

L<http://github.com/lestrrat/ZeroMQ-Perl>

=head1 AUTHOR

Daisuke Maki C<< <daisuke@endeworks.jp> >>

Steffen Mueller, C<< <smueller@cpan.org> >>

=head1 COPYRIGHT AND LICENSE

The ZeroMQ module is

Copyright (C) 2010 by Daisuke Maki

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.0 or,
at your option, any later version of Perl 5 you may have available.

=cut
